doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -11,92 +11,83 @@ These tactics use the equality :g:`eq:forall A:Type, A->A->Prop` defined in
+     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
+     the tactic fails.
 
-      Analogous to :n:`rewrite @term` but rewriting is done following
-      the clause :token:`goal_occurrences`. For instance:
+   :n:`@occurrences`
+     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
+     any of them can be rewritten.  If not specified, only the first occurrence
+     in the conclusion is replaced.
 
-      + :n:`rewrite H in H'` will rewrite `H` in the hypothesis
-        ``H'`` instead of the current goal.
-      + :n:`rewrite H in H' at 1, H'' at - 2 |- *` means
-        :n:`rewrite H; rewrite H in H' at 1; rewrite H in H'' at - 2.`
-        In particular a failure will happen if any of these three simpler tactics
-        fails.
-      + :n:`rewrite H in * |-` will do :n:`rewrite H in H'` for all hypotheses
-        :g:`H'` different from :g:`H`.
-        A success will happen as soon as at least one of these simpler tactics succeeds.
-      + :n:`rewrite H in *` is a combination of :n:`rewrite H` and :n:`rewrite H in * |-`
-        that succeeds if at least one of these two tactics succeeds.
+     .. todo PR above is correct?  rewrite ... at 1 2 3 works even if only 1 is rewritten?
 
-      Orientation :g:`->` or :g:`<-` can be inserted before the :token:`term` to rewrite.
+   :n:`@ltac_expr3`
+     If specified, is used to resolve all side conditions generated by the tactic.
 
-   .. tacv:: rewrite @term at @occurrences
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term_1`.
 
-      Rewrite only the given :token:`occurrences` of :token:`term`. Occurrences are
-      specified from left to right as for pattern (:tacn:`pattern`). The rewrite is
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+   :n:`@natural`
+     If followed by :n:`?`, gives the maximum number of rewrites that will be performed.
+     Otherwise, it specifies the exact number of rewrites to perform (whether followed by
+     `!` or not).  If the given number of rewrites can't be done, the tactic fails.
 
-   .. tacv:: rewrite @term by @tactic
+   :n:`?`
+     Performs the rewrite as many times as possible (possibly zero times). This form never fails.
 
-      Use tactic to completely solve the side-conditions arising from the
-      :tacn:`rewrite`.
+   :n:`!`
+     Performs the rewrite as many times as possible and at least once.  Otherwise this
+     form fails.
 
-   .. tacv:: rewrite {+, @orientation @term} {? in @ident }
+   .. todo PR: The following seems to be standard behavior for @occurrences, so I plan
+      to drop (most of?) it.
+      And I don't get the part about
+      "for all hypotheses :g:`H'` different from :g:`H`".  What do these mean?
```

Comment:
You could also consider this to be an example, that could be added as such (`.. example::`) to the description of `occurrences`. I believe that the important piece of info here is the part that you don't understand: if you rewrite with one of your hypothesis in all of your hypotheses, the tactic won't attempt to rewrite in the hypothesis you are rewriting with.

Reply:
I think we need to mention that `one_term` can be a lemma or the name of a hypothesis, which the current text doesn't make clear.  And maybe in several other tactics.

Is `rewrite (contrapositive P Q) by dec` (from `FSetDecide.v`) is a setoid rewrite?  I don't `eq` in Logic.Decidable.contrapositive.  So not sure what to say just below the `tacn`..

I added some words under `occurrences`

Reply:
Please respond to the previous comment.

Reply:
> I think we need to mention that `one_term` can be a lemma or the name of a hypothesis, which the current text doesn't make clear. And maybe in several other tactics.

I agree.

> Is `rewrite (contrapositive P Q) by dec` (from `FSetDecide.v`) is a setoid rewrite? I don't `eq` in Logic.Decidable.contrapositive. So not sure what to say just below the `tacn`.

Yeah, I think so. I'm not sure what determines whether setoid rewrite is used or not...